design/feature-flags.md

```diff
@@ -0,0 +1,98 @@
+# "Future" Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is
+the correct default behavior for the CDK. The problem, of course, is that
+breaking changes are only allowed in major versions and those are rare.
+
+This document describes a proposal for a pattern/mechanism called "future
+flags". It will allow us to introduce breaking behavior which is disabled by
+default (so existing projects will not be affected) but enabled automatically
+for new projects created through `cdk init`.
+
+> The name "**future flags**" is a wordplay on "feature flags" (dah!) to
+> represent features that are "from the future" and indicate that these specific
+> feature flags are only disabled by default only until the next release, after
+> which they will either be removed _or_ turned into regular feature flags that
+> people can select if they so desire.
+>
+> Having said all that, I am okay if we feel this is unnecessary and we can call
+> these just "feature flags" and not have a semantic distinction between the
+> ones that are "temporary" and permanent.
+
+## Approach
+
+The basic idea is that new breaking behavior will always be disabled by default
+and only enabled when a certain CDK context parameter is set. If not enabled,
+the system will continue to behave exactly like it used to without breaking any
+existing projects.
+
+When we release a new major version of the AWS CDK, we will flip this behavior
+or completely remove the legacy behavior.
+
+In order for new projects to pick up this new behavior automatically, we will
+modify `cdk init` to inject the set of "future flags" into the generated
+`cdk.json` file. This means that the new project will have the latest behavior,
+but projects that were created prior to the introduction of this feature will
+have the same legacy behavior based on the set of capabilities that were
+available at the time of the project's creation. This list will be cleaned up
+every time we release a major version of course.
+
+Using fine-grained flags will allow users of old projects to pick up specific
+new "future" behaviors by manually adding the specific keys to their `cdk.json`
+file, without risking breakage in other unexpected areas.
+
+## Alternative Considered
+ 
+We considered an alternative of "bundling" new capabilities under a single flag
+that specifies the CDK version which created the project, but this means that
+users won't have the ability to pick and choose which future capabilities they
+want to enable in case they need them but don't want to take the risk of
+unexpected changes.
+ 
+The downside of the fine-grained approach is that it could result in a "blowing
+up" new `cdk.json` files in case there will be many new breaking capabilities
+between major releases. But this is hypothetical and even if this list ends up
+with 20 features before we release the next major version, I still think the
+benefits outweigh the risks of the alternative approach.
```

Comment:
Can you articulate the risks you're afraid of?

Reply:
If I used version 1.56.0 to create my project and now it is 1.788.0 and I change my cdk.json to point to that version because there was a "breaking fix" there I want to opt into, I may end up with 20 different breaking changes without really knowing what they are. Granted, I can cherry-pick this feature and only enable it, but it will be very tempting to users to simply change the version in their `cdk.json` to the latest one and get hurt. I also think that this version number will be confusing. People will ask themselves - is this the CDK CLI version? The framework? Do I need to upgrade it every time I upgrade the framework? It going to be hard to explain. I think a granular set of flags is a very explicit mechanism. 

And as I metioned before, if we find ourselves with 100s of these, we can always introduce groups or introduce this version-based mehcniam, so I don't think we will be taking a one-way door by not introducing it in the first place. Do you?

Reply:
Added a new under a new section for "future development"